Handle Interrupted errors

jeromegn opened the pull request:
According to [various](https://github.com/tokio-rs/mio/blob/95393bc9aa3c4a38821b09405eb6217f56467ba3/examples/tcp_server.rs#L120) [sources](https://docs.rs/tokio/1.1.0/tokio/io/unix/struct.AsyncFd.html) and chats with developers on the tokio Discord, we should retry io errors Interrupted and WouldBlock immediately.

WouldBlock needs to clear the read/write state, but Interrupted should just be retried right away.

We've had issues with this in our VMs where our acceptor got interrupted and would not continue.

Tim-Zhang commented:
Hi @jeromegn, There are differences between tokio 0.2 and 1.0, We don't need to loop and retry in tokio 0.2 because [poll_read_ready()](https://docs.rs/tokio/0.2.25/src/tokio/io/poll_evented.rs.html#308) do that for us.

jeromegn commented:
You're right, we don't need to continue after `clear_read_ready` calls.

However, handling io errors of the `Interrupted` kind here is a good idea. Not doing so results in our listeners not accepting anymore.

Tim-Zhang commented on src/stream.rs:
```diff
@@ -116,6 +116,7 @@ impl VsockStream {
                 self.io.clear_write_ready(cx)?;
                 Poll::Pending
             }
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => self.poll_write_priv(cx, buf),
```
```suggestion
        loop {
            match self.io.get_ref().write(buf) {
                Err(e) if e.kind() == ErrorKind::WouldBlock => {
                    self.io.clear_write_ready(cx)?;
                    return Poll::Pending;
                }
                Err(e) if e.kind() == ErrorKind::Interrupted => continue,
                x => return Poll::Ready(x),
            }
        }
```

jeromegn replied:
Oh, you preferred the loop? I thought it was a bit more dangerous than just calling the same function.

Tim-Zhang replied:
Not only about loop, it's about what to retry, I think we need retry the `self.io.get_ref().write` not `poll_write_priv `.

Tim-Zhang commented:
@jeromegn thanks a lot. It is indeed a problem with `EINTR`

Please fix the commit message check, The commit body is required.

jeromegn commented:
This commit check is annoying for merges with master. It says it wants a body, but do I really need to explain a merge from master? :)

Tim-Zhang commented:
@jeromegn Ah, It would be better we use rebase instead of merge.

jeromegn commented:
Ah, I think we got it :)

Sorry for all the force pushing.

You probably want to review one last time. I made the changes and brought back the loop, but only for the inner poll (not the readiness poll).

Tim-Zhang commented:
@jeromegn great job, LGTM.

Can I merge this next week？I need some time for testing and I think this pr should be reviewed by more guys.

jeromegn commented:
Sure! I'm going to modify the branch we're using in production to use these changes and I'll report on my success.

jeromegn commented:
We've tested this in production over the weekend and everything seems smooth!

Tim-Zhang commented:
> We've tested this in production over the weekend and everything seems smooth!

Good news. I will test it in our project in these two days

shanecurran reviewed (approved):
LGTM

lifupan commented:
Hi @jeromegn 

Since the listener and stream socket had been set as nonblock mode, and it shouldn't be blocked and thus they shouldn't  get the interrupted error. I'm curious why it happened.

Tim-Zhang commented:
@lifupan good question. It' weird we never met this situation in our project. @jeromegn

jeromegn commented:
@Tim-Zhang @lifupan I have no idea why, but it happened consistently in our environment. We're running firecracker microvms.

We can't use the crate at all without this.

Tim-Zhang commented:
@jeromegn non-blocking socket should never fail with EINTR according to https://stackoverflow.com/questions/14134440/eintr-and-non-blocking-calls

I think we should look into it before merge. Can you provide a simple demo which does not work?

Tim-Zhang commented:
I would rather believe that there is a problem with the socket setting.
failed to set to nonblock mode? can you detect the flag use fcntl?

Maybe there are bugs in vsock_rs::set_nonblocking.

jeromegn commented:
I don't think I'd be able to provide a reproduction. vsocks are not a widely discussed subject, so when I looked into it, I couldn't figure out why this was happening. Our setup is complex and has many moving parts. We run arbitrary code from users, so anything can happen inside the VM. It's hard to provide a reproduction.

I suspect the stackoverflow answer is not completely right and that `EINTR` **can** happen since it's shown being handled in `mio` examples. I've also been advised to handle it from various `tokio` folks.

I don't know enough about vsocks and interrupt system calls to formulate a good argument for adding this code except that we cannot use this crate unless it includes it because it does in fact happen in our production environment.

It doesn't happen at first, but then after a few `accept`s, it does start happening. We're using the `Incoming` stream with `warp`. The latter will stop accepting as soon as `accept` returns an error (like `EINTR`). It's bound to happen to somebody else.

If this is not merged, we will be force to maintain a fork :(

> Maybe there are bugs in vsock_rs::set_nonblocking.

Maybe, but I don't think so. Until #7 was merged, accepting in a loop didn't work at all. Perhaps it doesn't work like it should, but it does at least makes it non-blocking.

Tim-Zhang reviewed (changes requested):
I think we should look into it.

jeromegn commented:
For what it's worth, this is all the code we're doing:

```rust
let incoming = VsockListener::bind(&SockAddr::new_vsock(3, 12345))?.incoming();
warp::serve(filters).serve_incoming(incoming)
```

jeromegn commented:
Reading the stackoverflow answer more carefully: the user only know about the EINTR and non-blocking interactions on macos/freebsd. They also are only talking theoretically. They aren't taling about vsocks either.

There is much that's different here!
- Firecracker microVM
- Linux
- Virtio

Any of which might cause this behaviour to be different. It could be a bug in firecracker, or a difference with how Linux does this. Like I said before, vsocks are not used all that much, at least they're not discussed publicly much. Before this crate came along, we had our own implementation, but I'd prefer using a crate where I can contribute back.

It's fine if you want to do more research, but I don't have more time to spend on this, unfortunately.

Tim-Zhang commented:
@lifupan @jeromegn 
After doing some research, I think there will be `EINTR` error in `epoll_wait of mio, and since mio 0.7 `poll` stop retrying when receive `EINTR`

Because We use mio 0.6 in our project, we haven't met this problem.

jeromegn commented:
Perhaps we only need to add the retry logic on `poll_accept`, which doesn't use mio except for `poll_read_ready` which does not throw.

If I understand correctly, mio 0.6 will handle EINTR on read and write because they use `Poll::poll`? However, mio 0.7 stopped handling it. Tokio added logic to handler that on read/write.

lifupan commented:
> I don't think I'd be able to provide a reproduction. vsocks are not a widely discussed subject, so when I looked into it, I couldn't figure out why this was happening. Our setup is complex and has many moving parts. We run arbitrary code from users, so anything can happen inside the VM. It's hard to provide a reproduction.
> 

Hi @jeromegn 

I think the right way to fix an issue is to find the root cause and fixed it, if you think it's not reproduced in the community and there's no convincing argument  to let other reviewers accept an workaround patch to be merged.

> I suspect the stackoverflow answer is not completely right and that `EINTR` **can** happen since it's shown being handled in `mio` examples. I've also been advised to handle it from various `tokio` folks.
> 

Actually the "mio" examples didn't deal with the `EINTR` in the accept procedure, and your case happened in the accept calling, thus the example couldn't help to confirm that it's needed to deal with the EINTR in poll_accept.

jeromegn commented:
> I think the right way to fix an issue is to find the root cause and fixed it

I agree.

I don't know what to add though. I've moved on ;) I will maintain a fork because I don't have time to make a reproduction for you. I can guarantee I can reproduce it with our complex setup, but recreating a minimal reproduction would be too time-consuming.

Tim-Zhang commented:
For controversial reasons, temporarily closed, waiting for new evidence to appear.

jeromegn commented:
For what it's worth, this issue was still happening for us and I have to maintain a fork with the fix which is to continue looping on EINTR errors (instead of breaking).

I don't have more evidence than: without the fix, the connection becomes unresponsive after some amount of time. We've been using this crate with over 15000 VMs and we kept getting odd vsock errors and timeouts from our host trying to communicate with the guests.

These are my changes btw: https://github.com/jeromegn/tokio-vsock/commit/be8244eccd67f761bfc900ab22af9675bce6608a

Now looking at how other libraries do it, it seems like perhaps in _some_ situations interrupted should be ignored and in others it should be handled.

Looking at `mio`, their `tcp_server` example effectively loops if it gets a EINTR error on either side:

https://github.com/tokio-rs/mio/blob/b2bbc1029849c939031fb396475d1a19da9786a6/examples/tcp_server.rs#L124-L126

https://github.com/tokio-rs/mio/blob/b2bbc1029849c939031fb396475d1a19da9786a6/examples/tcp_server.rs#L154

This is true of all their example code.

Tim-Zhang commented:
@jeromegn The TcpStream is in blocking mode in the example, if they set the mode non-blocking use [set_nonblocking](https://doc.rust-lang.org/stable/std/net/struct.TcpStream.html#method.set_nonblocking), then the interrupted could be ignored.

Only the slow syscall will be interrupted by a signal, non-blocking calls will return immediately and never be think as slow.

fasterthanlime commented:
Hey @Tim-Zhang & @lifupan — just catching up with this thread now, I was wondering what the status was on this pull request? Here's what I can make out from the conversation so far:

> Because We use mio 0.6 in our project, we haven't met this problem.

`tokio-vsock` depends on `tokio@1` right now, which definitely seems to pull in mio 0.7:

```
cargo tree -i mio --depth 2
mio v0.7.13
└── tokio v1.11.0
    ├── fly-init v0.1.0 (/home/amos/work/init)
    ├── h2 v0.3.4
    ├── hyper v0.14.12
    ├── netlink-proto v0.6.0
    ├── netlink-sys v0.6.0
    ├── rtnetlink v0.7.0
    ├── tokio-stream v0.1.7
    ├── tokio-tungstenite v0.13.0
    ├── tokio-util v0.6.8
    ├── tokio-vsock v0.3.1
    └── warp v0.3.1
```

> non-blocking socket should never fail with EINTR according to [stackoverflow.com/questions/14134440/eintr-and-non-blocking-calls](https://stackoverflow.com/questions/14134440/eintr-and-non-blocking-calls)

Seconding Jerome's read here, the comment starts with:

> I cannot give you a definitive answer to this question, and the answer may further vary from system to system [...]

The stackoverflow commenter then goes to check the the MacOS 10.8 kernel, which is very far from what I think either of us are running in production.

> Actually the "mio" examples didn't deal with the `EINTR` in the accept procedure, and your case happened in the accept calling, thus the example couldn't help to confirm that it's needed to deal with the EINTR in poll_accept.

I suppose a "run-once on a linux desktop system" example that accepts a single connection hasn't had the pleasure of running into this specific failure mode. That doesn't mean it doesn't exist!

---

@Tim-Zhang it seems you replied to this thread with "The TcpStream is in blocking mode in this example" (then deleted the comment).

I don't think that's true: mio's `TcpListener` ends up calling `new_socket`, which does set `SOCK_NONBLOCK`:

https://github.com/tokio-rs/mio/blob/a2a4d12f75c21f260ed7d9ebad175d298d0bf8ae/src/sys/unix/net.rs#L25

As @jeromegn mentioned, there's a lot of moving parts to our setup: I wouldn't be surprised if somewhere in the "firecracker microVMs / linux / virtio" stack there was a bug/inconsistency that causes some syscalls on non-blocking sockets to occasionally return `EINTR` when they technically never should. And retrying fixes it.

I understand being hesitant to merge something that fixes an issue you can't reproduce, but I'm wondering: what's the maintenance burden here? If you're never seeing EINTR in your codebase, that code will never get executed. For us, it's the difference between "15'000 VMs work fine" and "things get stuck".

Maintaining a fork is no fun at all. Can we work something out here?

Tim-Zhang commented:
@fasterthanlime I deleted the comment as soon as I posted it, you are not supposed to see it. and sorry for the wrong info.
